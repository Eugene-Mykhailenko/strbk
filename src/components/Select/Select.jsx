/* eslint-disable jsx-a11y/label-has-for */
import React, { Component } from 'react';
import PropTypes from 'prop-types';
import classNames from 'classnames';
import { CSSTransition } from 'react-transition-group';
import styles from './Select.scss';
import animationStyles from '../../assets/styles/base/animations.scss';
import Icon from '../Icon/Icon';

export default class Select extends Component {
  constructor(props) {
    super(props);

    this.state = {
      hasFocus: false,
    };

    this.internalOnChange = this.internalOnChange.bind(this);
  }

  getContainerClassNames() {
    const { hasError, isRequired, disabled, isDefault } = this.props;
    const { hasFocus } = this.state;

    return classNames(styles['form-group'], {
      [styles['has-focus']]: hasFocus,
      [styles['has-error']]: hasError,
      [styles['is-required']]: isRequired,
      [styles['is-disabled']]: disabled,
      [styles['is-default']]: isDefault,
    });
  }

  itemList() {
    const { items, isDefault } = this.props;

    /* eslint-disable */
    const itemList = items.map((item, index) => (
      <option key={index} value={item.value}>
        {item.label}
      </option>
    ));

    !isDefault &&
      itemList.unshift(<option key="placeholder" disabled={true} selected />);

    return itemList;
  }

  internalOnChange(event) {
    const { onChange, disabled } = this.props;

    if (disabled) {
      return;
    }

    if (event.target.value.length) {
      this.setState({
        hasFocus: true,
      });
    } else {
      this.setState({
        hasFocus: false,
      });
    }

    onChange(event);
  }

  renderLabel() {
    const { placeholder, isDefault } = this.props;

    return (
      placeholder &&
      !isDefault && <label className={styles.label}>{placeholder}</label>
    );
  }

  renderErrorText() {
    const { hasError, errorText, isDefault } = this.props;

    return (
      !isDefault && (
        <CSSTransition
          in={hasError}
          classNames={{
            enter: animationStyles['fade-in-start'],
            enterActive: animationStyles['fade-in-end'],
            exit: animationStyles['fade-out-start'],
            exitActive: animationStyles['fade-out-end'],
          }}
          timeout={300}
          unmountOnExit
        >
          <div className={styles['error-text']}>{errorText}</div>
        </CSSTransition>
      )
    );
  }

  render() {
    const { disabled, value, width } = this.props;

    return (
      <div
        className={this.getContainerClassNames()}
        style={{
          width,
        }}
      >
        {this.renderLabel()}
        <div className={styles['custom-select']}>
          <select
            disabled={disabled}
            onChange={this.internalOnChange}
            defaultValue={value}
          >
            {this.itemList()}
          </select>
          <Icon
            className={styles.arrow}
            glyph="UII_ChevronDown"
            width="16"
            height="16"
          />
        </div>
        {this.renderErrorText()}
      </div>
    );
  }
}

Select.propTypes = {
  /** List of select items (id, value, label) */
  items: PropTypes.arrayOf(
    PropTypes.shape({
      value: PropTypes.string,
      label: PropTypes.string,
    }).isRequired,
  ),
  /** Curntly selected value */
  value: PropTypes.string,
  /** Select Placeholder/Label */
  placeholder: PropTypes.string,
  /** Select width */
  width: PropTypes.string,
  /** Callback function for event */
  onChange: PropTypes.func.isRequired,
  /** Add star for label if this field is required */
  isRequired: PropTypes.bool,
  /** Error indication */
  hasError: PropTypes.bool,
  /** Error message */
  errorText: PropTypes.string,
  /** Disabled Select */
  disabled: PropTypes.bool,
  /** Default type */
  isDefault: PropTypes.bool,
};

Select.defaultProps = {
  items: [],
  value: undefined,
  placeholder: '',
  width: undefined,
  hasError: false,
  isRequired: false,
  errorText: undefined,
  disabled: false,
  isDefault: false,
};
