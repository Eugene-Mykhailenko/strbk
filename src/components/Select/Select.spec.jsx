import React from 'react';
import Select from './Select';

jest.mock('../Icon/importIcons');

const config = {
  items: [
    { value: '1', label: '1' },
    { value: '2', label: '2' },
    { value: '3', label: '3' },
    { value: '4', label: '4' },
    { value: '5', label: '5' },
  ],
  placeholder: 'Label',
  onChange: jest.fn(),
  isRequired: true,
  hasError: true,
  errorText: 'Error Text exemple',
  disabled: true,
  isDefault: true,
};

describe('Select', () => {
  it('should render floating label', () => {
    const select = shallow(
      <Select
        placeholder={config.placeholder}
        items={config.items}
        onChange={config.onChange}
      />,
    );

    expect(select.find('label').text()).toEqual(config.placeholder);
    expect(select).toMatchSnapshot();
  });

  it('should be able to render required state', () => {
    const select = shallow(
      <Select
        isRequired={config.isRequired}
        items={config.items}
        onChange={config.onChange}
      />,
    );

    expect(select.prop('className')).toMatch(/is-required/);
    expect(select).toMatchSnapshot();
  });

  it('should be able to render error state', () => {
    const select = shallow(
      <Select
        hasError={config.hasError}
        items={config.items}
        onChange={config.onChange}
      />,
    );

    expect(select.prop('className')).toMatch(/has-error/);
    expect(select).toMatchSnapshot();
  });

  it('should be able to render error message', () => {
    const select = shallow(
      <Select
        items={config.items}
        hasError={config.hasError}
        errorText={config.errorText}
        onChange={config.onChange}
      />,
    );

    expect(select.prop('className')).toMatch(/has-error/);
    expect(select.find('.error-text').text()).toEqual(`${config.errorText}`);
    expect(select).toMatchSnapshot();
  });

  it('should be able to render disabled state', () => {
    const select = shallow(
      <Select
        disabled={config.disabled}
        items={config.items}
        onChange={config.onChange}
      />,
    );

    expect(select.prop('className')).toMatch(/is-disabled/);
    expect(select.find('select').prop('disabled')).toEqual(true);
    expect(select).toMatchSnapshot();
  });

  it('should not trigger onChange handler in a disabled state', () => {
    const select = shallow(
      <Select
        disabled={config.disabled}
        items={config.items}
        onChange={config.onChange}
      />,
    );

    select.find('select').simulate('change', { target: { value: '123' } });

    expect(config.onChange).toHaveBeenCalledTimes(0);
    expect(select).toMatchSnapshot();
  });

  it('should trigger onChange handler', () => {
    const select = shallow(
      <Select items={config.items} onChange={config.onChange} />,
    );

    select.find('select').simulate('change', { target: { value: '123' } });

    expect(config.onChange).toHaveBeenCalledTimes(1);
    expect(select).toMatchSnapshot();
  });

  it('onChange event should style select', () => {
    const select = shallow(
      <Select
        isPasswordType={config.isPasswordType}
        items={config.items}
        onChange={config.onChange}
      />,
    );

    select.find('select').simulate('change', { target: { value: '123' } });

    expect(select.prop('className')).toMatch(/has-focus/);
    expect(select).toMatchSnapshot();
  });

  it('should be able to render default state', () => {
    const select = shallow(
      <Select
        items={config.items}
        onChange={config.onChange}
        isDefault={config.isDefault}
      />,
    );

    expect(select.prop('className')).toMatch(/is-default/);
    expect(select.find('label').length).toEqual(0);
    expect(select).toMatchSnapshot();
  });

  it('should not be able to render error message in default select', () => {
    const select = shallow(
      <Select
        items={config.items}
        hasError={config.hasError}
        errorText={config.errorText}
        onChange={config.onChange}
        isDefault={config.isDefault}
      />,
    );

    expect(select.find('.error-text').length).toEqual(0);
    expect(select).toMatchSnapshot();
  });
});
