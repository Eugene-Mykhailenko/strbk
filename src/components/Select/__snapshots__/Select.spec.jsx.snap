// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Select onChange event should style select 1`] = `
<div
  className="form-group has-focus"
  style={
    Object {
      "width": undefined,
    }
  }
>
  <div
    className="custom-select"
  >
    <select
      disabled={false}
      onChange={[Function]}
    >
      <option
        disabled={true}
        key="placeholder"
        selected={true}
      />
      <option
        key="0"
        value="1"
      >
        1
      </option>
      <option
        key="1"
        value="2"
      >
        2
      </option>
      <option
        key="2"
        value="3"
      >
        3
      </option>
      <option
        key="3"
        value="4"
      >
        4
      </option>
      <option
        key="4"
        value="5"
      >
        5
      </option>
    </select>
    <Icon
      className="arrow"
      fill="currentColor"
      glyph="UII_ChevronDown"
      height="16"
      width="16"
    />
  </div>
  <CSSTransition
    classNames={
      Object {
        "enter": "fade-in-start",
        "enterActive": "fade-in-end",
        "exit": "fade-out-start",
        "exitActive": "fade-out-end",
      }
    }
    in={false}
    timeout={300}
    unmountOnExit={true}
  >
    <div
      className="error-text"
    />
  </CSSTransition>
</div>
`;

exports[`Select should be able to render default state 1`] = `
<div
  className="form-group is-default"
  style={
    Object {
      "width": undefined,
    }
  }
>
  <div
    className="custom-select"
  >
    <select
      disabled={false}
      onChange={[Function]}
    >
      <option
        key="0"
        value="1"
      >
        1
      </option>
      <option
        key="1"
        value="2"
      >
        2
      </option>
      <option
        key="2"
        value="3"
      >
        3
      </option>
      <option
        key="3"
        value="4"
      >
        4
      </option>
      <option
        key="4"
        value="5"
      >
        5
      </option>
    </select>
    <Icon
      className="arrow"
      fill="currentColor"
      glyph="UII_ChevronDown"
      height="16"
      width="16"
    />
  </div>
</div>
`;

exports[`Select should be able to render disabled state 1`] = `
<div
  className="form-group is-disabled"
  style={
    Object {
      "width": undefined,
    }
  }
>
  <div
    className="custom-select"
  >
    <select
      disabled={true}
      onChange={[Function]}
    >
      <option
        disabled={true}
        key="placeholder"
        selected={true}
      />
      <option
        key="0"
        value="1"
      >
        1
      </option>
      <option
        key="1"
        value="2"
      >
        2
      </option>
      <option
        key="2"
        value="3"
      >
        3
      </option>
      <option
        key="3"
        value="4"
      >
        4
      </option>
      <option
        key="4"
        value="5"
      >
        5
      </option>
    </select>
    <Icon
      className="arrow"
      fill="currentColor"
      glyph="UII_ChevronDown"
      height="16"
      width="16"
    />
  </div>
  <CSSTransition
    classNames={
      Object {
        "enter": "fade-in-start",
        "enterActive": "fade-in-end",
        "exit": "fade-out-start",
        "exitActive": "fade-out-end",
      }
    }
    in={false}
    timeout={300}
    unmountOnExit={true}
  >
    <div
      className="error-text"
    />
  </CSSTransition>
</div>
`;

exports[`Select should be able to render error message 1`] = `
<div
  className="form-group has-error"
  style={
    Object {
      "width": undefined,
    }
  }
>
  <div
    className="custom-select"
  >
    <select
      disabled={false}
      onChange={[Function]}
    >
      <option
        disabled={true}
        key="placeholder"
        selected={true}
      />
      <option
        key="0"
        value="1"
      >
        1
      </option>
      <option
        key="1"
        value="2"
      >
        2
      </option>
      <option
        key="2"
        value="3"
      >
        3
      </option>
      <option
        key="3"
        value="4"
      >
        4
      </option>
      <option
        key="4"
        value="5"
      >
        5
      </option>
    </select>
    <Icon
      className="arrow"
      fill="currentColor"
      glyph="UII_ChevronDown"
      height="16"
      width="16"
    />
  </div>
  <CSSTransition
    classNames={
      Object {
        "enter": "fade-in-start",
        "enterActive": "fade-in-end",
        "exit": "fade-out-start",
        "exitActive": "fade-out-end",
      }
    }
    in={true}
    timeout={300}
    unmountOnExit={true}
  >
    <div
      className="error-text"
    >
      Error Text exemple
    </div>
  </CSSTransition>
</div>
`;

exports[`Select should be able to render error state 1`] = `
<div
  className="form-group has-error"
  style={
    Object {
      "width": undefined,
    }
  }
>
  <div
    className="custom-select"
  >
    <select
      disabled={false}
      onChange={[Function]}
    >
      <option
        disabled={true}
        key="placeholder"
        selected={true}
      />
      <option
        key="0"
        value="1"
      >
        1
      </option>
      <option
        key="1"
        value="2"
      >
        2
      </option>
      <option
        key="2"
        value="3"
      >
        3
      </option>
      <option
        key="3"
        value="4"
      >
        4
      </option>
      <option
        key="4"
        value="5"
      >
        5
      </option>
    </select>
    <Icon
      className="arrow"
      fill="currentColor"
      glyph="UII_ChevronDown"
      height="16"
      width="16"
    />
  </div>
  <CSSTransition
    classNames={
      Object {
        "enter": "fade-in-start",
        "enterActive": "fade-in-end",
        "exit": "fade-out-start",
        "exitActive": "fade-out-end",
      }
    }
    in={true}
    timeout={300}
    unmountOnExit={true}
  >
    <div
      className="error-text"
    />
  </CSSTransition>
</div>
`;

exports[`Select should be able to render required state 1`] = `
<div
  className="form-group is-required"
  style={
    Object {
      "width": undefined,
    }
  }
>
  <div
    className="custom-select"
  >
    <select
      disabled={false}
      onChange={[Function]}
    >
      <option
        disabled={true}
        key="placeholder"
        selected={true}
      />
      <option
        key="0"
        value="1"
      >
        1
      </option>
      <option
        key="1"
        value="2"
      >
        2
      </option>
      <option
        key="2"
        value="3"
      >
        3
      </option>
      <option
        key="3"
        value="4"
      >
        4
      </option>
      <option
        key="4"
        value="5"
      >
        5
      </option>
    </select>
    <Icon
      className="arrow"
      fill="currentColor"
      glyph="UII_ChevronDown"
      height="16"
      width="16"
    />
  </div>
  <CSSTransition
    classNames={
      Object {
        "enter": "fade-in-start",
        "enterActive": "fade-in-end",
        "exit": "fade-out-start",
        "exitActive": "fade-out-end",
      }
    }
    in={false}
    timeout={300}
    unmountOnExit={true}
  >
    <div
      className="error-text"
    />
  </CSSTransition>
</div>
`;

exports[`Select should not be able to render error message in default select 1`] = `
<div
  className="form-group has-error is-default"
  style={
    Object {
      "width": undefined,
    }
  }
>
  <div
    className="custom-select"
  >
    <select
      disabled={false}
      onChange={[Function]}
    >
      <option
        key="0"
        value="1"
      >
        1
      </option>
      <option
        key="1"
        value="2"
      >
        2
      </option>
      <option
        key="2"
        value="3"
      >
        3
      </option>
      <option
        key="3"
        value="4"
      >
        4
      </option>
      <option
        key="4"
        value="5"
      >
        5
      </option>
    </select>
    <Icon
      className="arrow"
      fill="currentColor"
      glyph="UII_ChevronDown"
      height="16"
      width="16"
    />
  </div>
</div>
`;

exports[`Select should not trigger onChange handler in a disabled state 1`] = `
<div
  className="form-group is-disabled"
  style={
    Object {
      "width": undefined,
    }
  }
>
  <div
    className="custom-select"
  >
    <select
      disabled={true}
      onChange={[Function]}
    >
      <option
        disabled={true}
        key="placeholder"
        selected={true}
      />
      <option
        key="0"
        value="1"
      >
        1
      </option>
      <option
        key="1"
        value="2"
      >
        2
      </option>
      <option
        key="2"
        value="3"
      >
        3
      </option>
      <option
        key="3"
        value="4"
      >
        4
      </option>
      <option
        key="4"
        value="5"
      >
        5
      </option>
    </select>
    <Icon
      className="arrow"
      fill="currentColor"
      glyph="UII_ChevronDown"
      height="16"
      width="16"
    />
  </div>
  <CSSTransition
    classNames={
      Object {
        "enter": "fade-in-start",
        "enterActive": "fade-in-end",
        "exit": "fade-out-start",
        "exitActive": "fade-out-end",
      }
    }
    in={false}
    timeout={300}
    unmountOnExit={true}
  >
    <div
      className="error-text"
    />
  </CSSTransition>
</div>
`;

exports[`Select should render floating label 1`] = `
<div
  className="form-group"
  style={
    Object {
      "width": undefined,
    }
  }
>
  <label
    className="label"
  >
    Label
  </label>
  <div
    className="custom-select"
  >
    <select
      disabled={false}
      onChange={[Function]}
    >
      <option
        disabled={true}
        key="placeholder"
        selected={true}
      />
      <option
        key="0"
        value="1"
      >
        1
      </option>
      <option
        key="1"
        value="2"
      >
        2
      </option>
      <option
        key="2"
        value="3"
      >
        3
      </option>
      <option
        key="3"
        value="4"
      >
        4
      </option>
      <option
        key="4"
        value="5"
      >
        5
      </option>
    </select>
    <Icon
      className="arrow"
      fill="currentColor"
      glyph="UII_ChevronDown"
      height="16"
      width="16"
    />
  </div>
  <CSSTransition
    classNames={
      Object {
        "enter": "fade-in-start",
        "enterActive": "fade-in-end",
        "exit": "fade-out-start",
        "exitActive": "fade-out-end",
      }
    }
    in={false}
    timeout={300}
    unmountOnExit={true}
  >
    <div
      className="error-text"
    />
  </CSSTransition>
</div>
`;

exports[`Select should trigger onChange handler 1`] = `
<div
  className="form-group has-focus"
  style={
    Object {
      "width": undefined,
    }
  }
>
  <div
    className="custom-select"
  >
    <select
      disabled={false}
      onChange={[Function]}
    >
      <option
        disabled={true}
        key="placeholder"
        selected={true}
      />
      <option
        key="0"
        value="1"
      >
        1
      </option>
      <option
        key="1"
        value="2"
      >
        2
      </option>
      <option
        key="2"
        value="3"
      >
        3
      </option>
      <option
        key="3"
        value="4"
      >
        4
      </option>
      <option
        key="4"
        value="5"
      >
        5
      </option>
    </select>
    <Icon
      className="arrow"
      fill="currentColor"
      glyph="UII_ChevronDown"
      height="16"
      width="16"
    />
  </div>
  <CSSTransition
    classNames={
      Object {
        "enter": "fade-in-start",
        "enterActive": "fade-in-end",
        "exit": "fade-out-start",
        "exitActive": "fade-out-end",
      }
    }
    in={false}
    timeout={300}
    unmountOnExit={true}
  >
    <div
      className="error-text"
    />
  </CSSTransition>
</div>
`;
